add markdown and html support to dataframe

pngwn opened the pull request:
Allows dataframe cells to contain markdown or html.

- markdown will be processed in the same way that markdown in the `Markdown` component is (using `MarkdownIt`)
- Supports both default values and generated values (via predict function)
- If the table is interactive, the 'editable' value is the resulting HTML. This feels like a niche usecase but it is supported. We could optionally disable editing for html + markdown fields. Allowing the original source markdown to be the editable value is technically possible but introduces a degree of complexity I did not want to introduce in this PR, it would require significant changes to the implementation.

Closes #1364.

cc @radames

pngwn commented:
Just to add, this can be tested via the `demo/blocks_outputs` demo. The fn is at the top, the dataframes are at the bottom (for predict fn + default value testing).

pngwn commented:
Another thing, I didn't update it here (although I can) but we don't need a new instance of `MarkdownIt` for each class instance of `Markdown`. That can be a static class variable.

freddyaboulton commented on gradio/components.py:
```diff
@@ -2550,10 +2565,28 @@ def __validate_headers(headers: List[str] | None, col_count: int):
 
+    @staticmethod
+    def __process_markdown(data: List[List[Any]], datatype: str | List[str]):
+        if (
```
@pngwn I think I figured out the bug.

In the demo `datatype` is `['markdown', 'markdown', 'html']`. That's always not equal to the string "markdown" so the `or` evaluates to True.

A more verbose but easier to understand implementation would be

```python
        if (
            isinstance(datatype, list) and "markdown" not in datatype
        ):
            return data
        if isinstance(datatype, str) and datatype != "markdown":
            return data
```

Also `isinstance` is better that `type(....) is ...` in my opinion

freddyaboulton replied:
It can render the markdown now

<img width="1344" alt="image" src="https://user-images.githubusercontent.com/41651716/176918894-ce424308-387f-4fca-86a3-38679dfe8d30.png">

pngwn replied:
Happy to change, google guides me.

freddyaboulton commented on gradio/components.py:
```diff
@@ -2521,15 +2528,23 @@ def postprocess(self, y):
             return y
         if isinstance(y, str):
             y = pd.read_csv(str)
```
Unrelated to your PR but this is a bug right? It should be `y = pd.read_csv(y)` right?

pngwn replied:
It has to be. I'll fix in this PR.

freddyaboulton commented on gradio/components.py:
```diff
@@ -2550,10 +2565,28 @@ def __validate_headers(headers: List[str] | None, col_count: int):
                 )
             )
 
+    @staticmethod
```
nitpick but I think a `classmethod` might be more appropriate here since we modify `md` in this method.

freddyaboulton replied:
Yes! Changes the api to this

```python
    @classmethod
    def __process_markdown(cls, data: List[List[Any]], datatype: str | List[str]):
```

But the usage in `preprocess` will be the same so no need to change anything there.

pngwn replied:
That's nice, cleaner way to access the class. ty!

freddyaboulton reviewed (approved):
@pngwn This looks good to me once we fix the `if` statement in `process_markdown` . Thank you!

pngwn commented:
As i discussed with @freddyaboulton in slack, I've switched to normalising the `datatype` upfront, so it is always a list of strings matching the column count. This simplifies the `process_markdown` logic significantly.

@freddyaboulton Feel free to take another glance + check that the previous example you tried works correctly.

I also fixed the bug you caught.

hibestil commented on gradio/components.py:
```diff
@@ -2356,6 +2356,8 @@ class Dataframe(Changeable, IOComponent):
     Demos: filter_records, matrix_transpose, tax_calculator
     """
 
+    md = None
```
IMO, It'd be better for understandability if we rename this variable to `markdown` or `has_markdown`. WDYT @pngwn ?

pngwn replied:
Good point, this is actually the markdown parser instance. I'll rename it accordingly.
